src/main/kotlin/com/exactpro/th2/common/utils/message/MessageBatcher.kt

```diff
@@ -75,10 +81,16 @@ abstract class Batcher<T>(
 
     abstract fun onMessage(message: T)
 
-    protected fun add(key: Any, message: RawMessage) = batches.getOrPut(key, ::Batch).add(message.toGroup())
-    protected fun add(key: Any, message: Message) = batches.getOrPut(key, ::Batch).add(message.toGroup())
-    protected fun add(key: Any, message: AnyMessage) = batches.getOrPut(key, ::Batch).add(message.toGroup())
-    protected fun add(key: Any, group: MessageGroup) = batches.getOrPut(key, ::Batch).add(group)
+    protected fun add(key: Any, message: RawMessage.Builder) = batches.getOrPut(key, ::Batch).add {
+        message.metadataBuilder.idBuilder.timestamp = Instant.now().toTimestamp()
+        message.toGroup()
+    }
```

Comment:
the `batches.getOrPut(key, ::Batch)` is executed thread-safe, but `.add { ... }` can be executed in different thread at the same time

Reply:
If you take a look at `add` method implementation you will see that the passed lambda is executed under a lock. So, there should not be any problems.
If I am wrong, please, add some details regarding your concerns

Reply:
I have double check in IDE, yes you right